pkg/app/master/command/images/handler.go

```diff
@@ -131,17 +133,17 @@ func OnCommand(
 	if cparams.TUI { // `images --tui`
 		initialTUI := InitialTUI(images, true)
 		tui.RunTUI(initialTUI, true)
-		return nil
 	} else if cparams.GlobalTUI { // `tui` -> `i`
-		return images
+		// TODO - create a central store for the lookup key.
+		// As this key needs to be the same on the sender and the receiver.
+		xc.Out.DataChannels["images"] <- images
```

Comment:
The business logic doesn't need to know that channels are involved. The Execution Context should hide it for the caller: `xc.Out.Data("images", images)`. And then the `Data()` method would send to the channel. Let's start with this enhancement first in the next PR.

More to consider for the follow up enhancements:

Either way, for each output method (e.g., xc.Out.State, xc.Out.Info, etc) we'll need to have a way to `subscribe`, so we can consume and display what's in those calls. For this having a set of subscription channels like the images data channel you have here makes sense. You specify what execution context output methods you want to consume and then you read from the provided channels.

In addition to the subscription channels passed to `NewExecutionContext()` we'll need some kind of channel-based plumbing internally in the Execution Context Output struct, so all calls to `xc.Out.WhateverMethod()` end up sending to an internal channel. The internal channel loop in the Execution Context Output would consume from those internal channels and it would be responsible for doing the requested I/O or it would pass the data to the subscription channels if they are provided for the corresponding output method.

Another thing. to consider:

We'll have more than one message to deal with especially when we process the `xc.Out.State` calls. This means that we'll need to have a loop consuming from the Execution Context channels. With Bubble Tea we'll need to make it work with their model Update method. The `httpit` app can be used as a potential reference for the design: https://github.com/gonetx/httpit/blob/main/pit/tui.go#L108 Here for the default message case it generates a `tick` and for us it can be a `receive one message from the internal channel(s)` function. What kind of other designs have you seen that handle external sources of data turning them into BubbleTea commands?

Reply:
Thank you for all of the feedback here.

Regarding your question of turning external data sources and turning them into BubbleTea commands, I really admire the pug implementation.

Their approach to using pubsub - https://github.com/leg100/pug/blob/master/internal/pubsub/broker.go#L55 - feels like a pattern to strive for.

I am not sure if you want to use generics in the same way they do for their event processing, but it is a potential path - https://github.com/leg100/pug/blob/master/internal/resource/event.go#L14